# Speech Emotion Recognition
## Milestone 3

### Extracting MFCC mean data for 3 datasets RAVDESS (English), EMOVO (Italian), EMO-DB (German)

In [1]:
# importing libraries
import os
import librosa
import numpy as np
import pandas as pd

In [2]:
def extract_mfcc_mean(dataframe, csv_file):
    df = dataframe
    df1 = pd.DataFrame()
    filepath = [os.path.join(d, x) for x in df['file']]
    for x in filepath:
        y, sr = librosa.load(x) # set file and  sample rate
        mfcc = np.mean(librosa.feature.mfcc(y=y, sr=sr).T, axis=0)
        mfcc = mfcc.reshape(1,-1) # flatten vector array
        mfcc_df = pd.DataFrame(mfcc) 
        df1 = pd.concat([df1,mfcc_df],ignore_index=True) # add to df

    # auto-increment column names
    df1.set_axis(['mfcc_'+ str((x)+1) for x in range(len(df1.columns))],axis=1,inplace=True)

    df_mfcc_mean = pd.concat([df, df1], axis=1) # concatenate label + all vectors to one df
    df_mfcc_mean = df_mfcc_mean.loc[(df_mfcc_mean['emotion'] == '01')|(df_mfcc_mean['emotion'] == '03')|(df_mfcc_mean['emotion'] == '04')].reset_index(drop=True)
    df_mfcc_mean.to_csv(csv_file, index = False) # df to csv
    
    return df_mfcc_mean

### RAVDESS (English) dataset

In [3]:
# set directory
d = 'datasets/ravdess'
df = pd.DataFrame()

# labeling emotions from file names
df['file'] = [x for x in os.listdir(d)] # create file column of filenames
df['emotion']=[x.split('-')[2] for x in df['file']] # read emotion code from filename
df['emotion_label'] = df['emotion'].map({'01':'neutral','02':'calm','03':'happy','04':'sad','05':'angry','06':'fearful','07':'disgust','08':'surprised'})
df['dataset'] = 'ravdess'
ravdess = extract_mfcc_mean(df, 'ravdess_mfcc_mean.csv') # calling function

In [4]:
ravdess

file emotion emotion_label  dataset      mfcc_1  \
0    03-01-01-01-02-02-06.wav      01       neutral  ravdess -710.797791   
1    03-01-01-01-02-02-12.wav      01       neutral  ravdess -676.566650   
2    03-01-03-01-01-02-06.wav      03         happy  ravdess -574.169189   
3    03-01-03-01-01-02-12.wav      03         happy  ravdess -600.397705   
4    03-01-04-01-01-02-02.wav      04           sad  ravdess -668.933594   
..                        ...     ...           ...      ...         ...   
475  03-01-03-02-02-02-23.wav      03         happy  ravdess -544.956482   
476  03-01-03-01-01-02-23.wav      03         happy  ravdess -583.496765   
477  03-01-01-01-01-02-03.wav      01       neutral  ravdess -618.714294   
478  03-01-01-01-01-02-17.wav      01       neutral  ravdess -646.492065   
479  03-01-01-01-02-02-23.wav      01       neutral  ravdess -624.276184   

        mfcc_2     mfcc_3     mfcc_4     mfcc_5     mfcc_6  ...    mfcc_11  \
0    53.439884 -15.190397   8.793198  -0.554165 -10.416554  ...  -8.738665   
1    52.879017   4.699556   6.543307   4.021244  -2.198579  ...  -8.157342   
2    42.048836 -50.638577   7.814414  -3.251384 -23.890253  ... -17.436363   
3    60.840874 -14.167917   8.259460   2.347058  -7.515831  ...  -4.999807   
4    55.743156 -10.204071   9.223741  -5.716846  -2.916350  ...  -5.563672   
..         ...        ...        ...        ...        ...  ...        ...   
475  36.885582 -10.316902   0.327900  -2.503531  -8.492240  ...  -3.168387   
476  43.580547  -7.453956   2.450277   3.951019  -4.914914  ...  -3.484834   
477  67.533989   4.918677  17.342468  10.384308   2.009631  ...   0.093037   
478  68.444695   6.175968  18.342121  13.930529   5.193265  ...   1.354195   
479  61.348080  -4.145511   6.582844  13.251948  -7.262891  ...  -0.452468   

      mfcc_12   mfcc_13   mfcc_14   mfcc_15   mfcc_16   mfcc_17   mfcc_18  \
0   -2.204105 -5.876750 -3.916210 -5.274551 -3.837144 -3.003580 -3.944510   
1   -2.291065 -7.150641 -2.611762 -0.946685 -9.005078 -5.436606 -2.201399   
2   -2.909546 -5.733043 -9.023292 -3.229995 -9.130769 -7.111090 -0.888517   
3    2.084753 -7.822709 -2.762776  0.491094 -5.751363 -5.363528  0.215265   
4   -3.409941 -5.798491  0.649065 -6.239035 -3.230912 -4.998880 -2.160484   
..        ...       ...       ...       ...       ...       ...       ...   
475 -6.539998 -5.048376 -5.871434 -7.170031 -0.723186 -8.807537 -3.281103   
476 -4.629651 -3.355734 -4.060095 -5.756019  2.690181 -8.130227 -3.584724   
477  0.736255 -0.844652  2.690057 -3.900963  4.055541 -0.883574  0.725936   
478  0.540084  5.578449  2.630299  0.326959  1.133311 -0.669118 -1.331768   
479 -4.171318 -2.808946 -0.704987 -5.979613  1.303016 -6.966359 -2.940805   

      mfcc_19   mfcc_20  
0   -5.300639 -4.574412  
1   -7.016598 -4.207880  
2   -3.120326 -4.910990  
3   -4.589694 -0.819311  
4   -5.506312 -2.222621  
..        ...       ...  
475 -4.249179 -3.099562  
476 -4.911337 -3.921757  
477  1.466697 -1.099582  
478 -0.509085  2.086693  
479 -4.308662 -2.780153  

[480 rows x 24 columns]

In [5]:
ravdess.describe()

mfcc_1      mfcc_2      mfcc_3      mfcc_4      mfcc_5      mfcc_6  \
count  480.000000  480.000000  480.000000  480.000000  480.000000  480.000000   
mean  -613.532532   52.068867   -5.485148    9.425587    2.170682   -3.807982   
std     90.643127   15.384360   13.015884    7.551806    7.963208    7.734591   
min   -839.052673    5.043634  -52.612705  -13.626986  -17.960556  -26.156446   
25%   -681.137802   42.757570  -12.198376    4.538647   -4.382643   -9.132449   
50%   -623.596191   52.924192   -5.033622    9.591923    2.452451   -3.310220   
75%   -543.628815   63.481635    2.895319   14.781893    8.539844    1.446442   
max   -349.667389   96.921036   27.706251   32.357601   21.693514   18.022451   

           mfcc_7      mfcc_8      mfcc_9     mfcc_10     mfcc_11     mfcc_12  \
count  480.000000  480.000000  480.000000  480.000000  480.000000  480.000000   
mean    -6.119145   -8.705525   -7.601641   -1.443693   -3.833814   -3.023298   
std      6.353483    5.277839    4.865547    4.103918    4.759276    3.948844   
min    -24.503761  -25.173374  -24.121395  -19.195724  -17.436363  -14.189649   
25%    -10.742883  -12.548539  -10.472805   -4.112330   -7.367750   -5.274882   
50%     -6.402218   -8.470064   -7.441238   -1.291091   -3.584869   -3.027760   
75%     -1.679153   -5.076747   -4.220887    1.182982    0.037159   -0.914826   
max     12.091780    6.050869    6.491067   11.830771    6.230079   14.833461   

          mfcc_13     mfcc_14     mfcc_15     mfcc_16     mfcc_17     mfcc_18  \
count  480.000000  480.000000  480.000000  480.000000  480.000000  480.000000   
mean    -1.985953   -0.862267   -2.169683   -0.754014   -3.328196   -0.354413   
std      4.354392    3.848622    3.831002    4.241569    3.480430    3.381152   
min    -14.429558   -9.023292  -11.287531  -11.856011  -11.687413   -8.171184   
25%     -5.106197   -3.526639   -4.999036   -4.076194   -5.557786   -2.746242   
50%     -1.909532   -0.663963   -2.148734   -0.228237   -3.721580   -0.790188   
75%      0.417340    1.277742    0.159671    2.022313   -1.242028    1.185913   
max     18.314598   19.319366   12.058558   12.824487   10.488097   17.855349   

          mfcc_19     mfcc_20  
count  480.000000  480.000000  
mean    -1.961425   -0.438350  
std      3.193268    3.510068  
min     -8.989244   -9.548208  
25%     -4.438488   -2.941966  
50%     -2.220307   -0.583504  
75%      0.147214    1.688945  
max     11.156430   16.513723

In [6]:
ravdess.emotion_label.value_counts() # emotions resume

happy      192
sad        192
neutral     96
Name: emotion_label, dtype: int64

### For the MFCC mean features in RAVDESS Dataset we have:
 - 480 Samples
 - 20 features

### EMOVO (Italian) dataset

In [7]:
# set directory
d = 'datasets/emovo'
df = pd.DataFrame()

# labeling emotions from file names
df['file'] = [x for x in os.listdir(d)] # create file column of filenames
df['emotion']=[x.split('-')[0] for x in df['file']] # read emotion code from filename
df['emotion_label'] = df['emotion'].map({'neu':'neutral','gio':'happy','tri':'sad'})
df['emotion'] = df['emotion'].map({'neu':'01','gio':'03','tri':'04'})
df['dataset'] = 'emovo'

emovo = extract_mfcc_mean(df, 'emovo_mfcc_mean.csv') # calling function

In [8]:
emovo

file emotion emotion_label dataset      mfcc_1     mfcc_2  \
0    gio-f3-l2.wav      03         happy   emovo -438.659210  90.945412   
1    gio-m2-n4.wav      03         happy   emovo -381.991455  74.694130   
2    gio-f2-d1.wav      03         happy   emovo -351.303619  58.933956   
3    neu-f1-l2.wav      01       neutral   emovo -350.790100  72.042770   
4    tri-m2-b3.wav      04           sad   emovo -477.750122  93.088966   
..             ...     ...           ...     ...         ...        ...   
247  neu-m1-d2.wav      01       neutral   emovo -211.501724  96.357712   
248  neu-f3-b1.wav      01       neutral   emovo -391.464722  83.541176   
249  neu-f3-n2.wav      01       neutral   emovo -375.670013  88.580002   
250  gio-m3-d2.wav      03         happy   emovo -412.716156  82.858101   
251  tri-f3-n3.wav      04           sad   emovo -582.647522  67.667603   

        mfcc_3     mfcc_4     mfcc_5     mfcc_6  ...    mfcc_11   mfcc_12  \
0     1.975548  24.228107   0.568954 -18.860289  ... -11.396211 -2.385496   
1     3.889835  33.676838  23.849306   5.745421  ...   2.826153  2.099944   
2   -16.786831   3.503576 -11.277497 -12.390900  ... -18.486914 -5.986807   
3   -22.839170   9.917839 -10.057104 -12.878907  ...  -3.928934  6.754178   
4    30.927399  32.732807   6.938892  -8.153327  ...   4.493856  1.766160   
..         ...        ...        ...        ...  ...        ...       ...   
247 -10.747402  20.243452  -8.120827  15.337584  ... -13.404488  3.503304   
248   1.230163  31.568951   0.287323 -27.597214  ... -11.560918  0.359207   
249  -8.817126  25.659487  -2.283285 -26.933544  ... -12.806755  0.392837   
250  -3.903650  24.818924   3.185204  -6.396571  ... -12.904155  6.233284   
251  29.744291  37.926479  13.329883   2.549489  ... -10.200152  0.638745   

       mfcc_13    mfcc_14   mfcc_15   mfcc_16    mfcc_17   mfcc_18    mfcc_19  \
0    -5.311081   9.859286 -8.860078 -5.895369  -4.106612  2.787873  -4.073847   
1   -12.002060   1.871373  2.418774 -0.736813  -7.120077  3.707809 -11.482510   
2    -4.962160   6.068850 -4.679415  6.225265   7.580145  6.270100   2.546465   
3    -5.710007   7.345354 -1.511691 -1.715024   1.082778  0.698588   0.326906   
4    -0.709033   7.813618 -5.045843  5.599194  -6.495215  6.528111  -3.384206   
..         ...        ...       ...       ...        ...       ...        ...   
247   2.068519   1.257875 -9.073780 -0.548378 -11.848409 -4.896562 -11.298735   
248  -5.845253   3.588434 -6.970005  0.480922  -8.447999  1.279859  -6.914462   
249 -10.146322   7.668211 -7.864522  1.818684  -3.950670 -3.421546  -3.461961   
250   8.444356 -10.947403  5.737343  3.001041 -11.962973  1.692518  -5.299089   
251  -8.791301   9.589749  0.876629 -1.886477  -1.776423  1.750491  -4.360761   

      mfcc_20  
0   -4.779741  
1    7.354778  
2    3.463073  
3    0.731022  
4    2.357965  
..        ...  
247  1.131935  
248 -0.845170  
249 -1.646635  
250  1.442764  
251  2.499637  

[252 rows x 24 columns]

In [9]:
emovo.describe()

mfcc_1      mfcc_2      mfcc_3      mfcc_4      mfcc_5      mfcc_6  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean  -415.976990   74.529221    4.724198   27.368845    2.979171   -5.465407   
std     86.630135   17.169088   17.264542   10.520486    9.356462   12.491570   
min   -615.825256   32.559795  -45.583450   -2.835639  -17.586416  -36.959934   
25%   -470.631477   60.946551   -7.897140   19.769197   -3.864180  -13.103015   
50%   -419.643616   73.688408    2.851362   27.549038    1.847422   -5.605342   
75%   -366.868042   87.566872   18.817440   34.321334    9.670426    3.754173   
max   -194.345566  118.114182   37.237602   60.561783   25.882263   27.175123   

           mfcc_7      mfcc_8      mfcc_9     mfcc_10     mfcc_11     mfcc_12  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean    -4.140191    0.386168   -5.455523    1.528324   -7.234572    1.256689   
std      9.630249    8.062167    6.571039    3.996466    6.458002    4.145839   
min    -23.468594  -20.749580  -23.605265  -12.792185  -22.026756   -9.828698   
25%    -11.200729   -4.052018  -10.331577   -0.963005  -12.671474   -1.620874   
50%     -3.695697   -0.792602   -5.453765    1.932234   -7.413760    1.539653   
75%      1.932033    4.195651   -0.282327    4.302591   -1.774028    3.892348   
max     18.046434   28.368593    9.017344    9.868066    7.383246   14.312927   

          mfcc_13     mfcc_14     mfcc_15     mfcc_16     mfcc_17     mfcc_18  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean    -5.108392    2.979333   -3.623354    0.534533   -5.463378    1.598443   
std      4.936340    5.265877    5.112710    4.445865    4.603814    3.486533   
min    -15.432897  -19.570900  -19.893888  -16.490234  -19.023598   -8.692899   
25%     -8.593730   -0.247453   -7.032946   -2.511937   -8.599472   -0.646925   
50%     -5.786935    3.000885   -3.457349    0.878550   -5.429691    1.910679   
75%     -2.087386    7.254625   -0.018326    3.702810   -2.246435    4.170029   
max      8.444356   13.889460    7.461708   12.350214    8.359444   10.700441   

          mfcc_19     mfcc_20  
count  252.000000  252.000000  
mean    -4.268633   -0.032625  
std      3.725714    3.075431  
min    -12.260484   -8.935989  
25%     -6.464149   -1.981238  
50%     -4.433923   -0.174708  
75%     -2.815264    1.778575  
max     14.854718   17.257656

In [10]:
emovo.emotion_label.value_counts() # emotions resume

happy      84
neutral    84
sad        84
Name: emotion_label, dtype: int64

### For the MFCC mean features in EMOVO Dataset we have:
 - 252 Samples
 - 20 features

### EMO-DB (German) dataset

In [11]:
# set directory
d = 'datasets/emo-db'
df = pd.DataFrame()

# labeling emotions from file names
df['file'] = [x for x in os.listdir(d)] # create file column of filenames
df['emotion']=[x[5:6] for x in df['file']] # read emotion code from filename
df['emotion_label'] = df['emotion'].map({'N':'neutral','F':'happy','T':'sad','W':'angry','A':'fearful','E':'disgust','L':'bored'})
df['emotion'] = df['emotion'].map({'N':'01','F':'03','T':'04'})
df['dataset'] = 'emo-db'

emodb = extract_mfcc_mean(df, 'emodb_mfcc_mean.csv') # calling function

In [12]:
emodb

file emotion emotion_label dataset      mfcc_1      mfcc_2  \
0    14a07Na.wav      01       neutral  emo-db -259.049316  112.883835   
1    09a01Nb.wav      01       neutral  emo-db -249.528473  130.260941   
2    16a01Fc.wav      03         happy  emo-db -248.512070   69.238434   
3    09b02Na.wav      01       neutral  emo-db -209.059097  127.694366   
4    03a05Nd.wav      01       neutral  emo-db -235.498398  119.978088   
..           ...     ...           ...     ...         ...         ...   
207  08b02Tc.wav      04           sad  emo-db -304.191925   95.711159   
208  14b01Fa.wav      03         happy  emo-db -292.771698   82.961174   
209  12b01Ta.wav      04           sad  emo-db -249.234879  126.467003   
210  10b10Fc.wav      03         happy  emo-db -228.652023  116.932655   
211  08b01Fd.wav      03         happy  emo-db -233.413818   89.562843   

        mfcc_3     mfcc_4     mfcc_5     mfcc_6  ...   mfcc_11    mfcc_12  \
0     4.888266  57.514965 -15.869048  24.584776  ...  5.893105 -10.459047   
1   -37.481228  49.762962 -16.348997  18.585035  ...  0.691470 -11.256852   
2   -77.745789  47.553116 -36.481800  13.982468  ... -3.707649  -6.134758   
3   -16.226055  49.830849 -18.476400  23.398840  ... -2.055866  -7.414509   
4   -22.230228  58.755493 -10.186797  18.521570  ...  3.915649  -7.750413   
..         ...        ...        ...        ...  ...       ...        ...   
207  10.241700  41.151539  -5.626107  29.074667  ...  7.212532   1.839870   
208 -33.988777  53.030106 -20.012482  13.105911  ... -6.505236  -7.183577   
209 -12.782749  67.222687 -12.460497  14.756647  ...  6.292495  -2.341092   
210 -36.605366  33.946796  -5.097609  14.264622  ...  7.473591 -13.649033   
211 -29.015064  62.969440 -10.156769  18.139563  ... -2.921295  -8.069416   

      mfcc_13   mfcc_14    mfcc_15    mfcc_16   mfcc_17    mfcc_18    mfcc_19  \
0    1.932762  1.733875   9.053558  -1.285743  1.805781  -6.128288  -3.999542   
1    1.793442 -0.124424   0.148461 -10.777856  3.035627  -8.732730  -3.558454   
2    3.070889 -9.262952  13.330701  -7.775200  9.681377  12.155363  10.658295   
3   -0.128468 -1.931731  -2.823451 -10.023691  1.555701  -8.086412   0.577282   
4    0.493142 -8.525086  10.016065  -5.194012  0.818325  -4.077713   1.055301   
..        ...       ...        ...        ...       ...        ...        ...   
207  5.419566  4.738638  10.436954  -6.026187  6.787625   1.309348   0.528138   
208 -4.157574 -1.697314   6.139741 -12.042064  1.982729  -3.996978  -5.298385   
209  4.311428 -5.796004  14.774328  -6.673213  3.943429   2.434836   2.137385   
210  0.341717 -5.524143   6.080994  -2.727769  1.662302  -4.028440  -0.376716   
211  3.104062 -5.462015   7.828046 -15.582324 -0.392959  -2.371913  -4.589195   

       mfcc_20  
0    -2.245823  
1     1.423128  
2    11.566079  
3    -1.865425  
4     1.690238  
..         ...  
207   1.007173  
208  -4.637581  
209   1.006930  
210   0.903696  
211  -5.343225  

[212 rows x 24 columns]

In [13]:
emodb.describe()

mfcc_1      mfcc_2      mfcc_3      mfcc_4      mfcc_5      mfcc_6  \
count  212.000000  212.000000  212.000000  212.000000  212.000000  212.000000   
mean  -252.886200  108.789062  -25.121717   50.969498  -13.740086   19.366644   
std     32.058369   22.839128   20.316757   11.468048   10.294621    9.391883   
min   -347.857147   38.398624  -94.376839   18.792612  -41.692127   -6.591328   
25%   -271.431671   95.333725  -36.057244   43.766535  -19.606182   12.702532   
50%   -250.509705  112.307335  -22.543912   51.408689  -13.595678   19.474437   
75%   -231.157379  125.854635  -10.297853   58.782621   -6.247774   25.937757   
max   -175.561371  155.437332   15.366862   80.544861    8.119290   43.462093   

           mfcc_7      mfcc_8      mfcc_9     mfcc_10     mfcc_11     mfcc_12  \
count  212.000000  212.000000  212.000000  212.000000  212.000000  212.000000   
mean   -10.289747    1.045927  -11.464961   -4.995627    1.120372   -5.650929   
std     10.012188    6.215829    6.410326    5.286606    4.827195    4.941826   
min    -30.688217  -16.389030  -30.115343  -18.862846  -11.916354  -17.920300   
25%    -18.921326   -3.300151  -14.623170   -8.204295   -2.579814   -9.458831   
50%    -10.203106    1.713708  -11.816683   -5.179976    1.182103   -6.151442   
75%     -3.445502    5.698821   -8.037802   -1.312514    4.872350   -1.782698   
max     16.869390   13.718910    3.914332    6.963143   11.725938    5.027358   

          mfcc_13     mfcc_14     mfcc_15     mfcc_16     mfcc_17     mfcc_18  \
count  212.000000  212.000000  212.000000  212.000000  212.000000  212.000000   
mean     1.042401   -5.396743    9.353661   -7.559699    1.778623   -1.379231   
std      3.878642    5.076703    4.715319    4.806412    4.842526    4.193532   
min     -7.812547  -16.564299   -9.071821  -20.022951   -8.138435   -9.869681   
25%     -1.318169   -9.164022    6.749441  -11.383344   -1.208521   -3.867594   
50%      0.928428   -5.792533    9.629509   -7.635250    1.552468   -2.093369   
75%      3.682184   -2.021757   12.117776   -3.945388    4.022878    0.659787   
max     12.655239    8.226622   21.919044    4.089287   23.264423   14.897428   

          mfcc_19     mfcc_20  
count  212.000000  212.000000  
mean    -0.494978    0.139358  
std      4.229020    4.741411  
min     -9.381336   -8.049773  
25%     -3.439213   -2.893024  
50%     -0.893662   -0.328823  
75%      1.493958    1.730112  
max     17.638950   17.623308

In [14]:
emodb.emotion_label.value_counts() # emotions_resume

neutral    79
happy      71
sad        62
Name: emotion_label, dtype: int64

### MFCC mean in EMO-DB Dataset:
 - 212 Samples
 - 20 features

### Mergin datasets

In [15]:
df = pd.concat([ravdess, emovo, emodb]).reset_index(drop=True)

In [16]:
df

file emotion emotion_label  dataset      mfcc_1  \
0    03-01-01-01-02-02-06.wav      01       neutral  ravdess -710.797791   
1    03-01-01-01-02-02-12.wav      01       neutral  ravdess -676.566650   
2    03-01-03-01-01-02-06.wav      03         happy  ravdess -574.169189   
3    03-01-03-01-01-02-12.wav      03         happy  ravdess -600.397705   
4    03-01-04-01-01-02-02.wav      04           sad  ravdess -668.933594   
..                        ...     ...           ...      ...         ...   
939               08b02Tc.wav      04           sad   emo-db -304.191925   
940               14b01Fa.wav      03         happy   emo-db -292.771698   
941               12b01Ta.wav      04           sad   emo-db -249.234879   
942               10b10Fc.wav      03         happy   emo-db -228.652023   
943               08b01Fd.wav      03         happy   emo-db -233.413818   

         mfcc_2     mfcc_3     mfcc_4     mfcc_5     mfcc_6  ...    mfcc_11  \
0     53.439884 -15.190397   8.793198  -0.554165 -10.416554  ...  -8.738665   
1     52.879017   4.699556   6.543307   4.021244  -2.198579  ...  -8.157342   
2     42.048836 -50.638577   7.814414  -3.251384 -23.890253  ... -17.436363   
3     60.840874 -14.167917   8.259460   2.347058  -7.515831  ...  -4.999807   
4     55.743156 -10.204071   9.223741  -5.716846  -2.916350  ...  -5.563672   
..          ...        ...        ...        ...        ...  ...        ...   
939   95.711159  10.241700  41.151539  -5.626107  29.074667  ...   7.212532   
940   82.961174 -33.988777  53.030106 -20.012482  13.105911  ...  -6.505236   
941  126.467003 -12.782749  67.222687 -12.460497  14.756647  ...   6.292495   
942  116.932655 -36.605366  33.946796  -5.097609  14.264622  ...   7.473591   
943   89.562843 -29.015064  62.969440 -10.156769  18.139563  ...  -2.921295   

       mfcc_12   mfcc_13   mfcc_14    mfcc_15    mfcc_16   mfcc_17   mfcc_18  \
0    -2.204105 -5.876750 -3.916210  -5.274551  -3.837144 -3.003580 -3.944510   
1    -2.291065 -7.150641 -2.611762  -0.946685  -9.005078 -5.436606 -2.201399   
2    -2.909546 -5.733043 -9.023292  -3.229995  -9.130769 -7.111090 -0.888517   
3     2.084753 -7.822709 -2.762776   0.491094  -5.751363 -5.363528  0.215265   
4    -3.409941 -5.798491  0.649065  -6.239035  -3.230912 -4.998880 -2.160484   
..         ...       ...       ...        ...        ...       ...       ...   
939   1.839870  5.419566  4.738638  10.436954  -6.026187  6.787625  1.309348   
940  -7.183577 -4.157574 -1.697314   6.139741 -12.042064  1.982729 -3.996978   
941  -2.341092  4.311428 -5.796004  14.774328  -6.673213  3.943429  2.434836   
942 -13.649033  0.341717 -5.524143   6.080994  -2.727769  1.662302 -4.028440   
943  -8.069416  3.104062 -5.462015   7.828046 -15.582324 -0.392959 -2.371913   

      mfcc_19   mfcc_20  
0   -5.300639 -4.574412  
1   -7.016598 -4.207880  
2   -3.120326 -4.910990  
3   -4.589694 -0.819311  
4   -5.506312 -2.222621  
..        ...       ...  
939  0.528138  1.007173  
940 -5.298385 -4.637581  
941  2.137385  1.006930  
942 -0.376716  0.903696  
943 -4.589195 -5.343225  

[944 rows x 24 columns]

In [17]:
df.describe()

mfcc_1      mfcc_2      mfcc_3      mfcc_4      mfcc_5      mfcc_6  \
count  944.000000  944.000000  944.000000  944.000000  944.000000  944.000000   
mean  -479.802673   70.802643   -7.169680   23.545303   -1.186673    0.954041   
std    167.834930   28.673172   19.221079   19.034777   11.181626   13.799658   
min   -839.052673    5.043634  -94.376839  -13.626986  -41.692127  -36.959934   
25%   -625.510117   50.997111  -16.475286    9.063701   -7.734466   -8.477011   
50%   -494.801453   65.616982   -6.096371   18.549925   -0.761496   -0.679988   
75%   -312.821518   87.965134    4.237127   36.567626    7.432541    8.537668   
max   -175.561371  155.437332   37.237602   80.544861   25.882263   43.462093   

           mfcc_7      mfcc_8      mfcc_9     mfcc_10     mfcc_11     mfcc_12  \
count  944.000000  944.000000  944.000000  944.000000  944.000000  944.000000   
mean    -6.527483   -4.088561   -7.896347   -1.447996   -3.629051   -2.470862   
std      8.509386    7.888164    6.100149    4.925088    6.032141    4.909384   
min    -30.688217  -25.173374  -30.115343  -19.195724  -22.026756  -17.920300   
25%    -11.955757   -9.466549  -11.532403   -4.626462   -7.839258   -5.516582   
50%     -6.569205   -4.530219   -8.048238   -1.151628   -3.315633   -2.551972   
75%     -0.484995    0.808138   -3.773738    1.827210    0.772832    0.803418   
max     18.046434   28.368593    9.017344   11.830771   11.725938   14.833461   

          mfcc_13     mfcc_14     mfcc_15     mfcc_16     mfcc_17     mfcc_18  \
count  944.000000  944.000000  944.000000  944.000000  944.000000  944.000000   
mean    -2.139389   -0.855091    0.030130   -1.938433   -2.751310   -0.063249   
std      4.911906    5.407179    6.705268    5.387102    4.877484    3.761360   
min    -15.432897  -19.570900  -19.893888  -20.022951  -19.023598   -9.869681   
25%     -5.712306   -4.379707   -4.696408   -5.684390   -5.825110   -2.761888   
50%     -2.029235   -0.618527   -1.191586   -1.384208   -3.205200   -0.368227   
75%      1.099734    2.077843    3.567648    1.911236   -0.033368    2.046358   
max     18.314598   19.319366   21.919044   12.824487   23.264423   17.855349   

          mfcc_19     mfcc_20  
count  944.000000  944.000000  
mean    -2.248003   -0.200303  
std      3.834035    3.724017  
min    -12.260484   -9.548208  
25%     -4.759507   -2.734192  
50%     -2.710395   -0.355310  
75%      0.081786    1.733388  
max     17.638950   17.623308

In [18]:
df.emotion_label.value_counts()

happy      347
sad        338
neutral    259
Name: emotion_label, dtype: int64

### MFCC mean in ALL Datasets:
 - 944 Samples
 - 20 features

In [19]:
df.to_csv('all_mfcc_mean.csv', index = False)

In [25]:
df.groupby(['dataset']).emotion_label.value_counts()

dataset  emotion_label
emo-db   neutral           79
         happy             71
         sad               62
emovo    happy             84
         neutral           84
         sad               84
ravdess  happy            192
         sad              192
         neutral           96
Name: emotion_label, dtype: int64